File: J9Ch/src/J_3_Binary_Tree_Divide_Conquer/Optional_11/_448_Inorder_Successor_in_Binary_Search_Tree_Medium.cpp
/**
 448
 Inorder Successor in Binary Search Tree
 */


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p) {
        TreeNode *successor = NULL;
        while (root != NULL && root->val != p->val) {
            if (root->val > p->val) {
                successor = root;
                root = root->left;
            } else {
                root = root->right;
            }
        }

        if (root == NULL) {
            return NULL;
        }

        if (root->right == NULL) {
            return successor;
        }

        root = root->right;
        while (root->left != NULL) {
            root = root->left;
        }

        return root;
     }
};

// 方法二

class Solution {
public:
  TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p) {
    // write your code here
    if (!root || !p) {
      return NULL;
    }

    TreeNode *ans = NULL, *cur = root;
    while (cur != p) {
      if (p->val < cur->val) {
        ans = cur;
        cur = cur->left;
      }
      if (p->val > cur->val) {
        cur = cur->right;
      }
    }

    if (p->right) {
      cur = p->right;
      while (cur->left) {
        cur = cur->left;
      }
      ans = cur;
    }

    return ans;
  }
};

File: J9Ch/src/J_3_Binary_Tree_Divide_Conquer/Optional_11/_448_Inorder_Successor_in_Binary_Search_Tree_Medium.java
package J_3_Binary_Tree_Divide_Conquer.Optional_11;
import java.util.*;
/**
 448
 Inorder Successor in Binary Search Tree


 * Created by tianhuizhu on 6/28/17.
 */
public class _448_Inorder_Successor_in_Binary_Search_Tree_Medium {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }
    public class Solution {
        public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
            TreeNode successor = null;
            while (root != null && root != p) {
                if (root.val > p.val) {
                    successor = root;
                    root = root.left;
                } else {
                    root = root.right;
                }
            }

            if (root == null) {
                return null;
            }

            if (root.right == null) {
                return successor;
            }

            root = root.right;
            while (root.left != null) {
                root = root.left;
            }

            return root;
        }
    }
}
